modules/seqexec/engine/src/main/scala/seqexec/engine/Engine.scala

```diff
@@ -263,38 +263,45 @@ class Engine[D, U](stateL: Lens[D, Engine.State]) {
       case Null                    => pure(SystemUpdate(se, EventResult.Ok))
     }
 
-    (ev match {
+    ev match {
       case EventUser(ue)   => handleUserEvent(ue)
-      case EventSystem(se) => handleSystemEvent(se).flatMap(x => userReact.applyOrElse(se, (_:SystemEvent) => unit).map(_ => x))
-    })
+      case EventSystem(se) => handleSystemEvent(se).flatMap(x =>
+        userReact.applyOrElse(se, (_:SystemEvent) =>  unit).map(_ => x))
+    }
   }
 
   /** Traverse a process with a stateful computation. */
   // input, stream of events
   // initalState: state
   // f takes an event and the current state, it produces a new state, a new value B and more actions
   @SuppressWarnings(Array("org.wartremover.warts.ImplicitParameter"))
-  def mapEvalState[A, S, B](input: Stream[IO, A], initialState: S, f: (A, S) => IO[(S, B, Stream[IO, A])])(implicit ec: ExecutionContext): Stream[IO, B] = {
+  def mapEvalState[A, S, B](input: Stream[IO, A],
+                            initialState: S, f: (A, S) => IO[(S, B, Option[Stream[IO, A]])])
+                           (implicit ec: ExecutionContext): Stream[IO, B] = {
     Stream.eval(fs2.async.unboundedQueue[IO, Stream[IO, A]]).flatMap { q =>
       Stream.eval_(q.enqueue1(input)) ++
         q.dequeue.joinUnbounded.evalMapAccumulate(initialState) { (s, a) =>
           f(a, s).flatMap {
-            case (ns, b, st) =>
-              q.enqueue1(st) >> IO.pure((ns, b))
+            case (ns, b, None)     => IO.pure((ns, b))
+            case (ns, b, Some(st)) => q.enqueue1(st) >> IO.pure((ns, b))
```

Comment:
It is a bit strange that you could change from `q.enqueue1(Option[A])` to `q.enqueue1(A)`. Am I missing something?

Reply:
Yes, you are missing something. Previously, `st` was a `Stream` that most of the time was `Stream.empty`. Now it is an `Option[Stream]`, where the case when it was `Stream.empty` was replaced by `None`.